### PR TITLE
Speed up element data

### LINKED ISSUE #638: Extended ElementData Sync
Trying a lot of different techniques to sync data between server and clients i'm always facing bandwidth and cpu drops on server-side or fps-drops on clientside.

Summarizing my experience, i would like to suggest a feature, something like

setElementDataEx(element, key, value, syncType=normal, syncTargets = root, forceSync=true)
forceElementDataSync()

This function should work in conjunction with the classic elementData, but the difference is how things will be synchronized.

syncType:
1)  only players listed in syncTargets will receive the update. if new player joins or reconnects the server, the date is hidden for him, but if same key and same value will be set targeting to this player, he should receive the update.

2) normal - direct update will receive only syncTargets, sets data as public, if new player join the server, he should receive this data as well.

3) stream - this data should be synchronized only when player streams the element or element already streamed in. syncTargets is ignored here, but can be used to filter players who should receive data when element streamed. When element streams out, the data should stay in memory, so next time only delta will be synced. (it's a little bit tricky, we need to store some timestamp of last sync, but it will save a bandwidth and cpu.)

forceSync:
if true - data should be directly sent like setElementData, if false - the MTA should push it into buffer and collect all changes until current scope ends or method forceElementDataSync() triggered. This means the data should be synchronized not like many single packets, but one batch operation, or can be splitted in to few buckets for better shipping and client-side processing. The order of operations should be keeped. If newly set value overlaps previously set one in pending buffer, it should be overwritten.

forceElementDataSync() - forces operations buffer to be synced.

p.s. i know that most of it can be done on lua side, but there is a few limitations with setElementData without sync, which forces you to keep a duplicated element data tree. Actually i already did most of it in lua, but some of operations causing FPS drops due to lua quickness, so i think it should be implemented closely to core :)

From [mantis](https://bugs.mtasa.com/view.php?id=9250).

### PR DESCRIPTION
**Pull request by Pirulax**
This PR improves element data perfomrnace by 22%(at least for now server-side), I'll implement it client-side as well tomorrow, hopefully.

[Test resource](https://github.com/multitheftauto/mtasa-blue/files/3136547/tester.zip)

### CODE REVIEW

**Comment by Pirulax** [comment]
1e6(1m) iters

|type |old |new             |            faster by|
|-----------------------------------------|-------|-------------|-------------|
| set - sync: true  \| Key: int             | 8705  | 7332        | 15,77254451 |
| set - sync: false \| Key: int            | 6003  | 4542        | 24,33783108 |
| set - sync: true  \| Key: 32byte key      | 5865  | 4115        | 29,83802217 |
| set - sync: false  \| Key: 32byte key     | 4447  | 2844        | 36,04677311 |
| get - inheritt:: true \| Key: int        | 2673  | 2418        | 9,539842873 |
| get - inheritt:: false \| Key: int       | 2690  | 2604        | 3,197026022 |
| get - inheritt:: true \| Key: 32byte key | 1252  | 817         | 34,74440895 |
| get - inheritt: false \| Key: 32byte key | 1282  | 873         | 31,90327613 |
|                                         |       |             |             |
| All time spent                          | 32917 | 25545       |             |
| Sped up by                              | 0%    | 22,39572257 |             |

File: Shared/sdk/SharedUtil.FastHashMap.h [hunk 1]
```diff
@@ -71,6 +71,13 @@ namespace SharedUtil
     // CFastHashMap helpers
     //
 
+    // Add a value for a key
+    template <class T, class V, class T2, class V2>
+    void MapInsert(CFastHashMap<T, V>& collection, const T2& key, const V2& value)
```
Comment (qaisjp) [inline Shared/sdk/SharedUtil.FastHashMap.h]: MapSet exists already but this is unused. It should removed if not being used.

File: Server/mods/deathmatch/logic/CElement.cpp [hunk 2]
```diff
@@ -690,36 +689,34 @@ bool CElement::GetCustomDataBool(const char* szName, bool& bOut, bool bInheritDa
...
-    CLuaArgument       oldVariable;
-    const SCustomData* pData = m_pCustomData->Get(szName);
-    if (pData)
-    {
-        oldVariable = pData->Variable;
+        return false;
     }
 
+    // We need the old variable as well
+    CLuaArgument OldVariable;
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CElement.cpp]:
```suggestion
    CLuaArgument oldVariable;
```

File: Server/mods/deathmatch/logic/CElement.cpp [hunk 3]
```diff
@@ -690,36 +689,34 @@ bool CElement::GetCustomDataBool(const char* szName, bool& bOut, bool bInheritDa
...
     // Set the new data
-    m_pCustomData->Set(szName, Variable, bSynchronized);
-
-    if (bTriggerEvent)
+    if (m_pCustomData->Set(szName, Variable, bSynchronized, &OldVariable))
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CElement.cpp]:
```suggestion
    if (m_pCustomData->Set(szName, Variable, bSynchronized, &oldVariable))
```

File: Server/mods/deathmatch/logic/CElement.cpp [hunk 4]
```diff
@@ -690,36 +689,34 @@ bool CElement::GetCustomDataBool(const char* szName, bool& bOut, bool bInheritDa
     return false;
 }
 
-void CElement::SetCustomData(const char* szName, const CLuaArgument& Variable, bool bSynchronized, CPlayer* pClient, bool bTriggerEvent)
+// Returns false in case we try to set the same value for the custom data and we didn't change the bSync, otherwise true
+bool CElement::SetCustomData(const char* szName, const CLuaArgument& Variable, bool bSynchronized, CPlayer* pClient, bool bTriggerEvent)
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CElement.cpp]:
```suggestion
bool CElement::SetCustomData(const char* szName, const CLuaArgument& variable, bool bSynchronized, CPlayer* pClient, bool bTriggerEvent)
```

File: Server/mods/deathmatch/logic/CElement.h [hunk 5]
```diff
@@ -140,7 +140,7 @@ class CElement
     bool           GetCustomDataInt(const char* szName, int& iOut, bool bInheritData);
     bool           GetCustomDataFloat(const char* szName, float& fOut, bool bInheritData);
     bool           GetCustomDataBool(const char* szName, bool& bOut, bool bInheritData);
-    void SetCustomData(const char* szName, const CLuaArgument& Variable, bool bSynchronized = true, CPlayer* pClient = NULL, bool bTriggerEvent = true);
+    bool SetCustomData(const char* szName, const CLuaArgument& Variable, bool bSynchronized = true, CPlayer* pClient = NULL, bool bTriggerEvent = true);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CElement.h:143]:
```suggestion
    bool SetCustomData(const char* szName, const CLuaArgument& variable, bool bSynchronized = true, CPlayer* pClient = NULL, bool bTriggerEvent = true);
```

Comment (Pirulax) [reply Server/mods/deathmatch/logic/CElement.h:143]: It was by default like that

File: Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp [hunk 6]
```diff
@@ -876,9 +876,7 @@ bool CStaticFunctionDefinitions::SetElementData(CElement* pElement, const char*
     assert(szName);
     assert(strlen(szName) <= MAX_CUSTOMDATA_NAME_LENGTH);
 
-    bool          bIsSynced;
-    CLuaArgument* pCurrentVariable = pElement->GetCustomData(szName, false, &bIsSynced);
-    if (!pCurrentVariable || *pCurrentVariable != Variable || bIsSynced != bSynchronize)
+    if (pElement->SetCustomData(szName, Variable, bSynchronize))
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:879]: this method needs modification to fix the capitalisation of variable (I'm aware it was like that before but it should be fixed nonetheless)

Comment (Pirulax) [reply Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:879]: Alright, will fix it tomororw, gotta sleep.

Should we keep the capitalisation of Variable in the sturct? Or I can change that to lowercase as well?

Comment (Pirulax) [reply Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Should we keep the capitalisation of Variable in the sturct?

Comment (Pirulax) [reply Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp]: Or I can change that to lowercase as well?

Comment (qaisjp) [reply Server/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:879]: actually, nevermind, we can address this another time. it was probably done that way for a reason although i can't think of one why

File: Server/mods/deathmatch/logic/CElement.cpp [hunk 7]
```diff
@@ -690,36 +689,34 @@ bool CElement::GetCustomDataBool(const char* szName, bool& bOut, bool bInheritDa
...
-        Arguments.PushString(szName);
-        Arguments.PushArgument(oldVariable);
-        Arguments.PushArgument(Variable);
-        CallEvent("onElementDataChange", Arguments, pClient);
+        if (bTriggerEvent)
+        {
+            // Trigger the onElementDataChange event on us
+            CLuaArguments Arguments;
+            Arguments.PushString(szName);
+            Arguments.PushArgument(OldVariable);
```
Comment (qaisjp) [inline Server/mods/deathmatch/logic/CElement.cpp]:
```suggestion
            Arguments.PushArgument(oldVariable);
```

**Comment by Einheit-101** [comment]
Amazing. Didnt expect a fix for that slow function anytime soon.

**Comment by Pirulax** [comment]
Well, this isn't all, I'll implement #638 as well.

**Comment by qaisjp** [comment]
> Well, this isn't all, I'll implement #638 as well.

Should be done in a separate pull request, but please wait until #638 has been approved before implementing it.

**Comment by qaisjp** [comment]
Please can you compare memory usage between versions?

**Comment by Pirulax** [comment]
I do.

| key type            | std::map    | CFastHashMap              
|------------------|-------------|-------------
| ~110byte          | 748mb            | 890mb         
| integer(0 - 1e6)| 748mb             | 366mb             

First code
```lua
local key = hash("sha512", "test"):sub(0, 110)
local data = hash("sha512", "datatest"):sub(0, 110)

local elem = Element("test")
for i = 0, 1e6 do
    elem:setData(key..i, data..i)
end
print("done")
```

Second code
```lua
local elem = Element("test")
for i = 0, 1e6 do
    elem:setData(i, i)
end
print("done")
```

**Comment by qaisjp** [comment]
Memory usage increase (assumed to be in mb) seems fine. Please can you measure against many multiple elements as well?

**Comment by Pirulax** [comment]
It is in MB, I'll add multiple elements later.
By the way, it can't handle more than 4m datas, at least on my machine, but that should be fine.
Maybe it's just up to my pc not having enough RAM, but on real servers that shouldn't be a problem.
And 4M ed's ain't real anyways

**Comment by Pirulax** [comment]
**Server(1e6 iters)**

|                                         | old  (ms)    | new (ms)         |speedup%|
|-----------------------------------------|----------|--------------|-------------|
| set - sync: true / Key: int             | 8705     | 7332         | 15,77254451 |
| set - sync: false / Key: int            | 6003     | 4542         | 24,33783108 |
| set - sync: true / Key: 32byte key      | 5865     | 4115         | 29,83802217 |
| set - sync: false / Key: 32byte key     | 4447     | 2844         | 36,04677311 |
| get - inheritt:: true / Key: int        | 2673     | 2418         | 9,539842873 |
| get - inheritt:: false / Key: int       | 2690     | 2604         | 3,197026022 |
| get - inheritt:: true / Key: 32byte key | 1252     | 817          | 34,74440895 |
| get - inheritt: false / Key: 32byte key | 1282     | 873          | 31,90327613 |
| **Total**                          | 32917    | 25545        |22,39572257|



**Memory usage(mb)**

|                                         | std::map | CFastHashMap |             | Difference | Increase(%) |
|----|------|--------|-------|------|------ |
| 1e6 entries(sync: true)                 | 748      | 890          |             | 142        | 15,95505618 |
| 10 * 1e5 entries(sync: true)            | 226      | 452          |             | 226        | 50          |



|| old                                     | new      |              |             |            |              
|-----------------------------------------|----------|--------------|-------------|------------|--------------|
| set - sync: true / Key: int             | 11762    | 7377         | 37,28107465 |            |              |
| set - sync: false / Key: int            | 9945     | 7095         | 28,65761689 |            |              |
| set - sync: true / Key: 32byte key      | 7265     | 5872         | 19,17412251 |            |              |
| set - sync: false / Key: 32byte key     | 7329     | 5128         | 30,03138218 |            |              |
| get - inheritt:: true / Key: int        | 2954     | 2442         | 17,3324306  |            |              |
| get - inheritt:: false / Key: int       | 3542     | 2484         | 29,87012987 |            |              |
| get - inheritt:: true / Key: 32byte key | 1436     | 915          | 36,28133705 |            |              |
| get - inheritt: false / Key: 32byte key | 1573     | 942          | 40,11443102 |            |              |
|                                         |          |              |             |            |              |
| All time spent                          | 45806    | 32255        |             |            |              |
| Sped up by                              | 0%       | 29,58346068  |             |            |              |
|                                         |          |              |             |            |              |
|                                         |          |              |             |            |              |
| Memory usage(mb)                        |          |              |             |            |              |
|                                         | std::map | CFastHashMap |             | Difference | Increase(%)  |
| 1e6 entries(sync: true)                 | 490      | 484          |             | -6         | -1,239669421 |
| 10 * 1e5 entries(sync: true)            | 437      | 634          |             | 197        | 31,07255521  |

**Comment by Pirulax** [comment]
image

**Comment by Pirulax** [comment]
Current bugs:
Getting CRPCFunctions FIX ME messages - Its not related to this PR, its the "intended behavior" as said by sbx.

Getting protocol error 60 - Not sure yet if its because this PR or not, but probably not.

**Comment by Pirulax** [comment]
Oh, i thought i didnt do anything in that commit
whatever

File: Client/mods/deathmatch/logic/rpc/CRPCFunctions.cpp [hunk 8]
```diff
@@ -53,7 +53,7 @@ SFixedArray<CRPCFunctions::SElementRPCHandler, CRPCFunctions::NUM_RPC_FUNCS> CRP
 
 CRPCFunctions::CRPCFunctions(CClientGame* pClientGame)
 {
-    m_bShowRPCs = true;
+    m_bShowRPCs = false;
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/rpc/CRPCFunctions.cpp]: i know this was an unintentional commit, but you can use the [`showrpcs`](https://github.com/multitheftauto/mtasa-blue/blob/e11685cab4beb7958ab202261f9c9d9b4ce71e58/Client/mods/deathmatch/CClient.cpp#L120) command for this.

Comment (Pirulax) [reply Client/mods/deathmatch/logic/rpc/CRPCFunctions.cpp]: it was originally false, I just changed it for debugging.
it was an intentional commit, I just didnt know theres something in it

**Comment by Pirulax** [comment]
Every bug that occurred in my tests(FIXME CRPCFuncs and protocol error 60) are in the current MTA release as well, so it wasnt my PR that introduced them.
I'll report those bugs now.

**Comment by Dutchman101** [comment]
Why did you close this PR, @Pirulax ?

**Comment by Lpsd** [comment]
@Dutchman101 the branch was renamed so a new PR was submitted #912

**Pull request closed without merging**